presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java

```diff
@@ -343,6 +366,16 @@ public void onSuccess(PagesResponse result)
 
                         if (token == HttpPageBufferClient.this.token) {
                             pages = result.getPages();
+                            if (pages.size() > pendingPages.size()) {
+                                throw new PageTransportErrorException(format("expected max number of pages to fetch: %s; found: %s", pendingPages.size(), pages.size()));
+                            }
+                            for (int i = 0; i < pages.size(); i++) {
+                                if (pages.get(i).getRetainedSizeInBytes() > pendingPages.get(i).toBytes()) {
+                                    // TODO: why after SerDe the size sometimes becomes smaller?
```

Comment:
Checked `PagesSerde`. When serializing and compressing a page on the server side, we construct the page with a `DynamicSliceOutput`. The retained size could be larger than the actual size unless we do a compression. A compression will make a copy of the dynamic slice with the actual length. After we transfer the page to the client side, the retained size is the actual size. Given this property, the client may reserve more memory than needed. This possibly won't be a big issue; the memory is still bounded by the buffer size. It is also impossible that a client reserves huge amount of memory and only fetches a page in a small size. Because if that is the case, the serve should compress the page. A compressed page should have the same retained size on both server and client.

Will add a comment.